File: sdk/python/flashq/worker.py
"""
flashQ Worker

Job processor with concurrent workers and graceful shutdown.
"""

import asyncio
import signal
from enum import Enum
from typing import Any, Callable, TypeVar, Generic, Awaitable

from .client import FlashQ
from .types import Job, ClientOptions, WorkerOptions, LogLevel
from .errors import FlashQError
from .utils.logger import Logger
from .constants import (
    WORKER_PULL_TIMEOUT,
    WORKER_ERROR_RETRY_DELAY,
    DEFAULT_WORKER_CONCURRENCY,
    DEFAULT_BATCH_SIZE,
    DEFAULT_CLOSE_TIMEOUT,
)

T = TypeVar("T")
R = TypeVar("R")


class WorkerState(str, Enum):
    """Worker state."""

    IDLE = "idle"
    STARTING = "starting"
    RUNNING = "running"
    STOPPING = "stopping"
    STOPPED = "stopped"


class WorkerEvents(Generic[T, R]):
    """
    Worker event callbacks with typed signatures.

    Type Parameters:
        T: Job data type
        R: Job result type
    """

    def __init__(self) -> None:
        self.on_ready: Callable[[], None | Awaitable[None]] | None = None
        self.on_active: Callable[[Job[T], int], None | Awaitable[None]] | None = None
        self.on_completed: Callable[[Job[T], R, int], None | Awaitable[None]] | None = None
        self.on_failed: Callable[[Job[T], Exception, int], None | Awaitable[None]] | None = None
        self.on_error: Callable[[Exception], None | Awaitable[None]] | None = None
        self.on_stopping: Callable[[], None | Awaitable[None]] | None = None
        self.on_stopped: Callable[[], None | Awaitable[None]] | None = None


class Worker(Generic[T, R]):
    """
    Job processor for flashQ queues.

    Supports concurrent processing, batch pulls, and graceful shutdown.

    Example:
        ```python
        async def process_email(job: Job[dict]) -> dict:
            # Process the job
            return {"sent": True}

        worker = Worker("emails", process_email)
        await worker.start()
        # Worker runs until stopped
        await worker.stop()
        ```
    """

    def __init__(
        self,
        queue: str | list[str],
        processor: Callable[[Job[T]], R | Awaitable[R]],
        client_options: ClientOptions | dict[str, Any] | None = None,
        worker_options: WorkerOptions | dict[str, Any] | None = None,
    ):
        # Queues
        if isinstance(queue, str):
            self._queues = [queue]
        else:
            self._queues = list(queue)

        self._processor = processor

        # Options
        if isinstance(client_options, dict):
            client_options = ClientOptions(**client_options)
        self._client_options = client_options or ClientOptions()

        if isinstance(worker_options, dict):
            worker_options = WorkerOptions(**worker_options)
        self._worker_options = worker_options or WorkerOptions()

        # State
        self._state = WorkerState.IDLE
        self._clients: list[FlashQ] = []
        self._workers: list[asyncio.Task[None]] = []
        self._processing = 0
        self._processed = 0
        self._failed_count = 0
        self._shutdown_event = asyncio.Event()

        # Events
        self.events: WorkerEvents[T, R] = WorkerEvents()

        # Logger
        self._logger = Logger(
            level=self._worker_options.log_level,
            prefix="Worker",
        )

        # Auto-start
        if self._worker_options.auto_start:
            asyncio.create_task(self.start())

    @property
    def state(self) -> WorkerState:
        """Get worker state."""
        return self._state

    @property
    def processing(self) -> int:
        """Get number of jobs currently processing."""
        return self._processing

    @property
    def processed(self) -> int:
        """Get total number of jobs processed."""
        return self._processed

    @property
    def failed(self) -> int:
        """Get total number of jobs failed."""
        return self._failed_count

    async def start(self) -> None:
        """Start the worker."""
        if self._state != WorkerState.IDLE:
            return

        self._state = WorkerState.STARTING
        self._logger.info("Starting", queues=self._queues)

        # Create clients (one per concurrency level)
        concurrency = self._worker_options.concurrency or DEFAULT_WORKER_CONCURRENCY

        for _ in range(concurrency):
            client = FlashQ(self._client_options)
            await client.connect()
            self._clients.append(client)

        # Start worker tasks
        for i in range(concurrency):
            task = asyncio.create_task(self._worker_loop(i))
            self._workers.append(task)

        self._state = WorkerState.RUNNING
        self._logger.info("Started", concurrency=concurrency)

        if self.events.on_ready:
            self._emit(self.events.on_ready)

    async def stop(self, force: bool = False) -> None:
        """
        Stop the worker.

        Args:
            force: If True, abort immediately without waiting for jobs
        """
        if self._state in (WorkerState.STOPPING, WorkerState.STOPPED):
            return

        self._state = WorkerState.STOPPING
        self._logger.info("Stopping", force=force)

        if self.events.on_stopping:
            self._emit(self.events.on_stopping)

        # Signal shutdown
        self._shutdown_event.set()

        if force:
            # Cancel all workers
            for task in self._workers:
                task.cancel()
        else:
            # Wait for workers with timeout
            timeout = self._worker_options.close_timeout or DEFAULT_CLOSE_TIMEOUT

            try:
                await asyncio.wait_for(
                    asyncio.gather(*self._workers, return_exceptions=True),
                    timeout=timeout / 1000,
                )
            except asyncio.TimeoutError:
                self._logger.warn("Timeout waiting for workers, forcing stop")
                for task in self._workers:
                    task.cancel()

        self._workers.clear()

        # Close clients
        for client in self._clients:
            await client.close()
        self._clients.clear()

        self._state = WorkerState.STOPPED
        self._logger.info("Stopped", processed=self._processed, failed=self._failed_count)

        if self.events.on_stopped:
            self._emit(self.events.on_stopped)

    async def wait(self) -> None:
        """Wait for worker to stop."""
        await self._shutdown_event.wait()

    async def _worker_loop(self, worker_id: int) -> None:
        """Main worker loop for processing jobs."""
        client = self._clients[worker_id]
        batch_size = self._worker_options.batch_size or DEFAULT_BATCH_SIZE
        queue_index = 0

        while not self._shutdown_event.is_set():
            try:
                # Round-robin through queues
                queue = self._queues[queue_index % len(self._queues)]
                queue_index += 1

                # Pull batch of jobs
                jobs = await client.pull_batch(queue, batch_size, WORKER_PULL_TIMEOUT)

                if not jobs:
                    continue

                # Process jobs concurrently
                tasks = [self._process_job(client, job, worker_id) for job in jobs]
                await asyncio.gather(*tasks, return_exceptions=True)

            except asyncio.CancelledError:
                break
            except Exception as e:
                self._logger.error("Worker loop error", error=str(e))

                if self.events.on_error:
                    self._emit(self.events.on_error, e)

                await asyncio.sleep(WORKER_ERROR_RETRY_DELAY / 1000)

    async def _process_job(self, client: FlashQ, job: Job[T], worker_id: int) -> None:
        """Process a single job."""
        self._processing += 1

        if self.events.on_active:
            self._emit(self.events.on_active, job, worker_id)

        try:
            # Call processor
            result = self._processor(job)
            if asyncio.iscoroutine(result):
                result = await result

            # Acknowledge success
            await client.ack(job.id, result)

            self._processed += 1

            if self.events.on_completed:
                self._emit(self.events.on_completed, job, result, worker_id)

        except Exception as e:
            self._failed_count += 1
            error_msg = str(e)

            try:
                await client.fail(job.id, error_msg)
            except Exception as fail_error:
                self._logger.error(
                    "Failed to report job failure",
                    job_id=job.id,
                    error=str(fail_error),
                )

            if self.events.on_failed:
                self._emit(self.events.on_failed, job, e, worker_id)

        finally:
            self._processing -= 1

    def _emit(self, callback: Callable[..., Any], *args: Any) -> None:
        """Emit event callback."""
        try:
            result = callback(*args)
            if asyncio.iscoroutine(result):
                asyncio.create_task(result)
        except Exception as e:
            self._logger.error("Event callback error", error=str(e))

    def on(self, event: str, callback: Callable[..., Any]) -> "Worker[T, R]":
        """
        Register event callback.

        Events:
            - ready: Worker started
            - active: Job started processing
            - completed: Job completed
            - failed: Job failed
            - error: Worker error
            - stopping: Worker stopping
            - stopped: Worker stopped
        """
        if event == "ready":
            self.events.on_ready = callback
        elif event == "active":
            self.events.on_active = callback
        elif event == "completed":
            self.events.on_completed = callback
        elif event == "failed":
            self.events.on_failed = callback
        elif event == "error":
            self.events.on_error = callback
        elif event == "stopping":
            self.events.on_stopping = callback
        elif event == "stopped":
            self.events.on_stopped = callback
        return self


def create_worker(
    queue: str | list[str],
    processor: Callable[[Job[T]], R | Awaitable[R]],
    client_options: ClientOptions | dict[str, Any] | None = None,
    worker_options: WorkerOptions | dict[str, Any] | None = None,
) -> Worker[T, R]:
    """
    Create a new worker.

    Convenience function for creating workers.
    """
    return Worker(queue, processor, client_options, worker_options)

File: sdk/go/flashq/worker.go
package flashq

import (
	"context"
	"fmt"
	"log"
	"sync"
	"sync/atomic"
	"time"
)

// JobProcessor is a function that processes a job and returns a result.
type JobProcessor func(job *Job) (interface{}, error)

// WorkerEventHandler handles worker events.
// Events that involve jobs include the workerID (processor goroutine index).
type WorkerEventHandler struct {
	OnReady     func()
	OnActive    func(job *Job, workerID int)
	OnCompleted func(job *Job, result interface{}, workerID int)
	OnFailed    func(job *Job, err error, workerID int)
	OnError     func(err error)
	OnStopping  func()
	OnStopped   func()
}

// Worker processes jobs from one or more queues with high performance.
type Worker struct {
	queues     []string
	processor  JobProcessor
	clientOpts ClientOptions
	workerOpts WorkerOptions
	events     WorkerEventHandler

	state      atomic.Int32
	client     *Client // Single shared client with pool
	wg         sync.WaitGroup
	ctx        context.Context
	cancel     context.CancelFunc
	processing atomic.Int32
	processed  atomic.Int64
	failed     atomic.Int64
	jobChan    chan *Job // Channel for job distribution
	logger     *log.Logger
}

// NewWorker creates a new high-performance worker.
func NewWorker(queues []string, processor JobProcessor, clientOpts *ClientOptions, workerOpts *WorkerOptions) *Worker {
	co := DefaultClientOptions()
	if clientOpts != nil {
		co = *clientOpts
	}
	wo := DefaultWorkerOptions()
	if workerOpts != nil {
		wo = *workerOpts
	}
	// Larger pool for workers
	if co.PoolSize < wo.Concurrency/2 {
		co.PoolSize = wo.Concurrency/2 + 1
	}
	ctx, cancel := context.WithCancel(context.Background())
	return &Worker{
		queues: queues, processor: processor, clientOpts: co, workerOpts: wo,
		ctx: ctx, cancel: cancel, logger: log.Default(),
	}
}

// NewWorkerSingle creates a worker for a single queue.
func NewWorkerSingle(queue string, processor JobProcessor, clientOpts *ClientOptions, workerOpts *WorkerOptions) *Worker {
	return NewWorker([]string{queue}, processor, clientOpts, workerOpts)
}

// State returns the current worker state.
func (w *Worker) State() WorkerState { return WorkerState(w.state.Load()) }

// Processing returns jobs currently being processed.
func (w *Worker) Processing() int { return int(w.processing.Load()) }

// Processed returns total jobs processed.
func (w *Worker) Processed() int64 { return w.processed.Load() }

// Failed returns total failed jobs.
func (w *Worker) Failed() int64 { return w.failed.Load() }

// On registers event handlers.
// For active/completed/failed events, handler signature includes workerID.
func (w *Worker) On(event string, handler interface{}) *Worker {
	switch event {
	case "ready":
		if h, ok := handler.(func()); ok { w.events.OnReady = h }
	case "active":
		if h, ok := handler.(func(job *Job, workerID int)); ok { w.events.OnActive = h }
	case "completed":
		if h, ok := handler.(func(job *Job, result interface{}, workerID int)); ok { w.events.OnCompleted = h }
	case "failed":
		if h, ok := handler.(func(job *Job, err error, workerID int)); ok { w.events.OnFailed = h }
	case "error":
		if h, ok := handler.(func(err error)); ok { w.events.OnError = h }
	case "stopping":
		if h, ok := handler.(func()); ok { w.events.OnStopping = h }
	case "stopped":
		if h, ok := handler.(func()); ok { w.events.OnStopped = h }
	}
	return w
}

// Start starts the worker.
func (w *Worker) Start(ctx context.Context) error {
	if w.State() != WorkerStateIdle { return nil }
	w.state.Store(int32(WorkerStateStarting))
	w.logInfo("Starting worker", "queues", w.queues)

	concurrency := w.workerOpts.Concurrency
	if concurrency < 1 { concurrency = 1 }

	// Single shared client with larger pool
	w.client = NewWithOptions(w.clientOpts)
	if err := w.client.Connect(ctx); err != nil {
		w.state.Store(int32(WorkerStateIdle))
		return fmt.Errorf("failed to connect: %w", err)
	}

	// Job channel for distribution
	w.jobChan = make(chan *Job, concurrency*2)

	// Start puller goroutines (fewer than workers)
	pullers := len(w.queues)
	if pullers > 4 { pullers = 4 }
	for i := 0; i < pullers; i++ {
		w.wg.Add(1)
		go w.pullerLoop(i)
	}

	// Start processor goroutines
	for i := 0; i < concurrency; i++ {
		w.wg.Add(1)
		go w.processorLoop(i)
	}

	w.state.Store(int32(WorkerStateRunning))
	w.logInfo("Worker started", "concurrency", concurrency)
	if w.events.OnReady != nil { w.events.OnReady() }
	return nil
}

// pullerLoop pulls jobs and distributes to processors.
func (w *Worker) pullerLoop(id int) {
	defer w.wg.Done()
	batchSize := w.workerOpts.BatchSize
	if batchSize < 1 { batchSize = 100 }
	pullTimeout := 5 * time.Second
	queueIdx := id

	for {
		select {
		case <-w.ctx.Done():
			return
		default:
		}
		queue := w.queues[queueIdx%len(w.queues)]
		queueIdx++

		jobs, err := w.client.PullBatch(queue, batchSize, pullTimeout)
		if err != nil {
			if w.events.OnError != nil { w.events.OnError(err) }
			select {
			case <-w.ctx.Done(): return
			case <-time.After(50 * time.Millisecond):
			}
			continue
		}
		for _, job := range jobs {
			select {
			case <-w.ctx.Done(): return
			case w.jobChan <- job:
			}
		}
	}
}

// processorLoop processes jobs from channel.
func (w *Worker) processorLoop(workerID int) {
	defer w.wg.Done()
	for {
		select {
		case <-w.ctx.Done():
			return
		case job := <-w.jobChan:
			if job == nil { return }
			w.processJob(job, workerID)
		}
	}
}

func (w *Worker) processJob(job *Job, workerID int) {
	w.processing.Add(1)
	defer w.processing.Add(-1)

	if w.events.OnActive != nil { w.events.OnActive(job, workerID) }

	result, err := w.processor(job)

	if err != nil {
		w.failed.Add(1)
		w.client.Fail(job.ID, err.Error())
		if w.events.OnFailed != nil { w.events.OnFailed(job, err, workerID) }
		return
	}

	w.client.Ack(job.ID, result)
	w.processed.Add(1)
	if w.events.OnCompleted != nil { w.events.OnCompleted(job, result, workerID) }
}

// Stop stops the worker gracefully.
func (w *Worker) Stop(ctx context.Context) error { return w.stop(ctx, false) }

// ForceStop stops the worker immediately.
func (w *Worker) ForceStop() error { return w.stop(context.Background(), true) }

func (w *Worker) stop(ctx context.Context, force bool) error {
	state := w.State()
	if state == WorkerStateStopping || state == WorkerStateStopped { return nil }

	w.state.Store(int32(WorkerStateStopping))
	w.logInfo("Stopping worker", "force", force)
	if w.events.OnStopping != nil { w.events.OnStopping() }

	w.cancel()
	close(w.jobChan)

	if !force {
		done := make(chan struct{})
		go func() { w.wg.Wait(); close(done) }()
		select {
		case <-done:
		case <-ctx.Done(): w.logWarn("Timeout waiting for workers")
		case <-time.After(w.workerOpts.CloseTimeout): w.logWarn("Timeout waiting for workers")
		}
	}

	if w.client != nil { w.client.Close() }
	w.state.Store(int32(WorkerStateStopped))
	w.logInfo("Worker stopped", "processed", w.Processed(), "failed", w.Failed())
	if w.events.OnStopped != nil { w.events.OnStopped() }
	return nil
}

// Wait blocks until the worker stops.
func (w *Worker) Wait() { w.wg.Wait() }

func (w *Worker) logInfo(msg string, args ...interface{}) {
	if w.workerOpts.LogLevel <= LogLevelInfo { w.logger.Printf("[INFO] Worker: %s %v\n", msg, args) }
}
func (w *Worker) logWarn(msg string, args ...interface{}) {
	if w.workerOpts.LogLevel <= LogLevelWarn { w.logger.Printf("[WARN] Worker: %s %v\n", msg, args) }
}
func (w *Worker) logError(msg string, args ...interface{}) {
	if w.workerOpts.LogLevel <= LogLevelError { w.logger.Printf("[ERROR] Worker: %s %v\n", msg, args) }
}
